src/main/java/kr/codesqaud/cafe/account/JoinFormValidator.java

```
@@ -23,8 +23,15 @@ public boolean supports(Class<?> clazz) {
     @Override
     public void validate(Object target, Errors errors) {
         JoinForm joinForm = (JoinForm) target;
+        if (errors.hasErrors()) {
+            return;
+        }
         if (userService.containsEmail(joinForm.getEmail())) {
             errors.rejectValue(EMAIL, "error.email.duplicate");
+            return;
+        }
+        if (!joinForm.getPassword().equals(joinForm.getReconfirmPassword())) {
+            errors.rejectValue("reconfirmPassword", "error.password.missMatch");
```

Validator를 구현하셨는데 `ConstraintValidator` 를 사용하지 않고 `Validator`를 사용한 이유가 궁금합니다.

---

`(!joinForm.getPassword().equals(joinForm.getReconfirmPassword()))`

해당 조건문에서 디미터 법칙을 사용하여 객체 그래프를 한번만 타는 연습을 하는건 어떻게 생각하세요?

키워드: 디미터 법칙

---

> Validator를 구현하셨는데 `ConstraintValidator` 를 사용하지 않고 `Validator`를 사용한 이유가 궁금합니다.

- ConstrainValidator를 모르고 있습니다. 찾아보고 한번 학습해보겠습니다.
- Validator는 온라인 강의에서 언급된 Spring Validator이여서 사용했습니다.

---

> `(!joinForm.getPassword().equals(joinForm.getReconfirmPassword()))`
> 
> 해당 조건문에서 디미터 법칙을 사용하여 객체 그래프를 한번만 타는 연습을 하는건 어떻게 생각하세요?
> 
> 키워드: 디미터 법칙

- 디미터 법칙 확인하고 적용하겠습니다. 감사합니다.🙇🏻‍♂️

---

> Validator를 구현하셨는데 ConstraintValidator 를 사용하지 않고 Validator를 사용한 이유가 궁금합니다.

- ConstraintValidator는 Javax에 포함한 Validator입니다.
- CoustraintValidator를 사용하여 Custom Valid Annoatation를 사용할 수 있는 편리함 있습니다.
---
저는 모든 기타 조건이 만족하는 경우 DB에서 email 존재 여부를 확인하고 싶었습니다.  그리고 모든 실패에 다른 메시지를 보내고 싶습니다.

1 `@ReportAsSingleViolation` 로 사용해봤는데 Message는 하나만 가능하는 것을 확인했습니다.

>  https://stackoverflow.com/questions/52299412/javax-annotation-constrainst-reportassingleviolation
      
2  Fail fast Mode를 해결을 시도했습니다. 그런데 모든 필드를 중 하나라도 실패하면 나머지는 확인하지 않아 목적과 다릅니다.
> https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/#section-fail-fast

---
CoustraintValidator으로 제가 원하는 로직을 구현하지 못하여 사용을 일단 보류하겠습니다.